celery/worker/request.py

```diff
@@ -93,7 +93,8 @@ def __init__(self, message, on_ack=noop,
                  maybe_make_aware=maybe_make_aware,
                  maybe_iso8601=maybe_iso8601, **opts):
         self._message = message
-        self._request_dict = message.headers if headers is None else headers
+        self._request_dict = (message.headers.copy() if headers is None
```

Comment:
Why are we copying the headers?

Reply:
`message.headers` is included in `message.properties` as `application_headers`; copying the headers here avoids creating a self-referential object. There's no reason (that I can recall, at least) that it absolutely cannot be self-referential, but it did lead to serialization errors in my test case. I think this is the least surprising way to handle this for end-users.

Reply:
it would great if you can also add unit test for the changes